### DISCUSSION TITLE
AllChem.GenerateDepictionMatching2DStructure not working for me despite common substructure!

### DISCUSSION THREAD

**Discussion by cagfa1** [rdkit/rdkit#4108] [Q&A] [answered]
I am trying to create molecular images of a series of chemicals that are all elaborations of an initial fragment. I have been trying to use the AllChem.GenerateDepictionMatching2DStructure with the initial fragment acting as a template to align the structures. There is a common substructure to all the compounds as rdFMCS.FindMCS(compounds) works fine. But no diagrams are produced when I run the GenererateDepiction code....

I used the following code:

```
from rdkit import Chem
from rdkit.Chem import Draw, AllChem
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline

Bound_Compounds = pd.read_csv('/biggin/b150/worc5014/docking/chemistry_sheets/structure_info_final.csv')
Bound_Products = []
for x in range(len(Bound_Compounds)):
    if Bound_Compounds.loc[x, 'Material'] == 'Product':
        Bound_Products.append(Bound_Compounds.loc[x, 'Smile'])
        
template = Chem.MolFromSmiles('NC(=O)N1CCN(CC1)C(=O)C2=CC=CO2') #Fragment SMILE

AllChem.Compute2DCoords(template)
ms = [Chem.MolFromSmiles(smi) for smi in Bound_Products]
for mol in ms:
    AllChem.GenerateDepictionMatching2DStructure(mol,template)
```

And get the following ERROR: 
```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-121-5ff53cae966e> in <module>
     17 ms = [Chem.MolFromSmiles(smi) for smi in Bound_Products]
     18 for mol in ms:
---> 19     AllChem.GenerateDepictionMatching2DStructure(mol,template)

ValueError: Depict error: Substructure match with reference not found.
```
I have repeated the process with: a) a truncated reference SMILE; and b) using a common substructure search (`rdFMCS.FindMCS(ms)`) of the elaborated fragments to create a SMART of shared substructure. In both cases there is no error but no diagrams are produced. I am not sure what this means - is the code working or is it failing to find substructures? How can I fix this?

The codes are shown below:
a)
```
Bound_Compounds = pd.read_csv('/biggin/b150/worc5014/docking/chemistry_sheets/structure_info_final.csv')
Bound_Products = []
for x in range(len(Bound_Compounds)):
    if Bound_Compounds.loc[x, 'Material'] == 'Product':
        Bound_Products.append(Bound_Compounds.loc[x, 'Smile'])   
        
ms = [Chem.MolFromSmiles(smi) for smi in Bound_Products] 

template1 = Chem.MolFromSmarts('[#6]-[#7]-[#6](=[#8])-[#7]1-[#6]-[#6]-[#7](-[#6]-[#6]-1)-[#6](=[#8])') #Common substructure SMART
AllChem.Compute2DCoords(template1)

for mol in ms:
    AllChem.GenerateDepictionMatching2DStructure(mol, template1)
```
b)
```
Bound_Compounds = pd.read_csv('/biggin/b150/worc5014/docking/chemistry_sheets/structure_info_final.csv')
Bound_Products = []
for x in range(len(Bound_Compounds)):
    if Bound_Compounds.loc[x, 'Material'] == 'Product':
        Bound_Products.append(Bound_Compounds.loc[x, 'Smile'])
        
template = Chem.MolFromSmiles('NC(=O)N1CCN(CC1)') #Truncated Fragment SMILE

AllChem.Compute2DCoords(template)
ms = [Chem.MolFromSmiles(smi) for smi in Bound_Products]
for mol in ms:
    AllChem.GenerateDepictionMatching2DStructure(mol,template)
```
Many thanks!

**Comment by apahl** [comment, marked as answer]
Without seeing the concrete examples, I am not sure what is going on in the cases where you get an error in the `GenerateDepictionMatching2DStructure`.

But for the cases, where you do not get error, you are just missing the function to actually draw the result, because `GenerateDepictionMatching2DStructure` is only calculating the coordinates, it is not drawing the molecules:

```Python
template1 = Chem.MolFromSmarts('[#6]-[#7]-[#6](=[#8])-[#7]1-[#6]-[#6]-[#7](-[#6]-[#6]-1)-[#6](=[#8])') #Common substructure SMART
AllChem.Compute2DCoords(template1)

for mol in ms:
    AllChem.GenerateDepictionMatching2DStructure(mol, template1)

Draw.MolsToGridImage(ms)
```

BTW, the statement to get the Smiles from the bound compounds where the material is "Product" could be shortened to e.g.:

`Bound_Products = Bound_Compounds.query("Material == 'Product'")["Smile"].values`

or

`Bound_Products = Bound_Compounds[Bound_Compounds["Material"] == "Product"]["Smile"].values`

KR Axel
